File: src/utils/schedulerHelpers.ts
// src/utils/schedulerHelpers.ts
import { Telegraf } from 'telegraf';
import { getTasksForDailyReminder } from '../services/taskManager';
import { getRandomPhraseForScheduler } from '../services/moeHandler';
import { Task } from '../types';

export async function sendDailyReminders(bot: Telegraf<any>) {
  console.log(`[${new Date().toISOString()}] Ejecutando recordatorios diarios...`);
  const pendingTasks: Task[] = await getTasksForDailyReminder();

  if (!pendingTasks.length) {
    console.log("No hay tareas pendientes para recordar.");
    return;
  }

  const remindersSentToChats = new Set<number>();

  for (const task of pendingTasks) {
    // Evitar spam si hay muchas tareas en un mismo chat para el mismo recordatorio
    if (remindersSentToChats.has(task.chat_id) && pendingTasks.filter(t => t.chat_id === task.chat_id).length > 3) {
        continue;
    }

    const userMention = `[${task.assigned_to_name}](tg://user?id=${task.assigned_to_id})`;
    const completeCommand = `/completar_${task.id}`;
    const message =
        `📢 **Recordatorio de Tarea Pendiente**\n\n` +
        `🔔 Tarea: ${task.description}\n` +
        `👤 Asignada a: ${userMention}\n` +
        `🆔 ID Tarea: \`${task.id}\`\n\n` +
        `¡No olvides completarla! Usa \`${completeCommand}\` cuando termines.`;
    try {
      // Enviar al chat donde se asignó la tarea
      if (task.chat_id) {
        await bot.telegram.sendMessage(task.chat_id, message, { parse_mode: 'MarkdownV2' });
        remindersSentToChats.add(task.chat_id);
        console.log(`Recordatorio enviado para tarea ${task.id} a chat ${task.chat_id}`);
        await new Promise(resolve => setTimeout(resolve, 300)); // Pequeña pausa
      } else {
        console.warn(`No se pudo enviar recordatorio para tarea ${task.id}: chat_id faltante.`);
      }
    } catch (error: any) {
      console.error(`Error al enviar recordatorio para tarea ${task.id}:`, error.message);
    }
  }
}

export async function sendRandomFunMessage(bot: Telegraf<any>) {
  const targetGroupId = parseInt(process.env.TELEGRAM_TARGET_GROUP_ID || '0');
  if (!targetGroupId) {
    console.warn("TELEGRAM_TARGET_GROUP_ID no configurado. No se enviará mensaje aleatorio.");
    return;
  }
  try {
    const messageToSend = getRandomPhraseForScheduler();
    await bot.telegram.sendMessage(targetGroupId, messageToSend);
    console.log(`Mensaje aleatorio enviado al grupo ${targetGroupId}: "${messageToSend}"`);
  } catch (error: any) {
    console.error("Error al enviar mensaje aleatorio al grupo:", error.message);
  }
}

File: app/utils/scheduler_helpers.py
# app/utils/scheduler_helpers.py
import os
from aiogram import Bot # Solo necesitas la instancia de Bot para enviar mensajes
from app.services import task_manager, moe_handler
from app.services.task_manager import Task # Importar el tipo Task

# Esta función será llamada por un cron job (ej: Vercel Cron)
async def trigger_daily_reminders(bot_instance: Bot):
    print(f"[{datetime.utcnow().isoformat()}] Ejecutando recordatorios diarios (scheduler)...")
    pending_tasks: List[Task] = await task_manager.get_tasks_for_daily_reminder()

    if not pending_tasks:
        print("Scheduler: No hay tareas pendientes para recordar.")
        return

    # ... (lógica de envío similar a la de Telegraf, adaptada a Aiogram)
    for task in pending_tasks:
        user_mention = f"<a href='tg://user?id={task['assigned_to_id']}'>{task['assigned_to_name']}</a>"
        complete_command = f"/completar_{task['id']}"
        message_text = (
            f"📢 <b>Recordatorio de Tarea Pendiente</b>\n\n"
            f"🔔 Tarea: {task['description']}\n"
            f"👤 Asignada a: {user_mention}\n"
            f"🆔 ID Tarea: <code>{task['id']}</code>\n\n"
            f"¡No olvides completarla! Usa <code>{complete_command}</code> cuando termines."
        )
        try:
            if task['chat_id']:
                await bot_instance.send_message(
                    chat_id=task['chat_id'],
                    text=message_text,
                    parse_mode="HTML" # Usar HTML consistente
                )
                print(f"Scheduler: Recordatorio enviado para tarea {task['id']} a chat {task['chat_id']}")
                # Pequeña pausa para evitar rate limits si hay muchos mensajes
                await asyncio.sleep(0.3)
        except Exception as e:
            print(f"Scheduler: Error al enviar recordatorio para tarea {task['id']}: {e}")

async def trigger_random_fun_message(bot_instance: Bot):
    target_group_id_str = os.getenv("TELEGRAM_TARGET_GROUP_ID")
    if not target_group_id_str:
        print("Scheduler: TELEGRAM_TARGET_GROUP_ID no configurado. No se enviará mensaje aleatorio.")
        return
    
    target_group_id = int(target_group_id_str)
    print(f"[{datetime.utcnow().isoformat()}] Enviando mensaje aleatorio (scheduler)...")
    try:
        message_to_send = moe_handler.get_random_phrase_for_scheduler()
        await bot_instance.send_message(
            chat_id=target_group_id,
            text=message_to_send,
            parse_mode="HTML" # Asume que la frase ya está formateada o es segura para HTML
        )
        print(f"Scheduler: Mensaje aleatorio enviado al grupo {target_group_id}")
    except Exception as e:
        print(f"Scheduler: Error al enviar mensaje aleatorio al grupo: {e}")

import asyncio # para asyncio.sleep
from datetime import datetime # para logging
from typing import List # para type hint